Sources/UIKit/PortalCarouselView.swift

```diff
@@ -0,0 +1,89 @@
+    where CustomComponentRendererType.MessageType == MessageType {
+    
+    public var isSnapToCellEnabled: Bool = false
+    
+    fileprivate let scrolledMessages: [MessageType?]
+    fileprivate var lastOffset: CGFloat = 0
+    fileprivate var selected: Int = 0 { didSet {
+            scrolledMessages[selected] |> { mailbox.dispatch(message: $0) }
+        }
+    }
+    
+    public override init(items: [CollectionItemProperties<MessageType>], customComponentRenderer: CustomComponentRendererType, layoutEngine: LayoutEngine, layout: UICollectionViewLayout) {
+        scrolledMessages = items.map { _ in .none }
+        super.init(items: items, customComponentRenderer: customComponentRenderer, layoutEngine: layoutEngine, layout: layout)
+    }
+    
+    public init(items: ZipList<CarouselItemProperties<MessageType>>?, customComponentRenderer: CustomComponentRendererType, layoutEngine: LayoutEngine, layout: UICollectionViewLayout) {
+        if let items = items {
+            let transform = { (item: CarouselItemProperties) -> CollectionItemProperties<MessageType> in
+                return collectionItem(
+                    onTap: item.onTap,
+                    identifier: item.identifier,
+                    renderer: item.renderer)
+            }
+            selected = Int(items.centerIndex)
+            scrolledMessages = items.map { $0.onScrolled }
+            super.init(items: items.map(transform), customComponentRenderer: customComponentRenderer, layoutEngine: layoutEngine, layout: layout)
+            scrollToItem(self.selected, animated: false)
+        } else {
+            scrolledMessages = []
+            super.init(items: [], customComponentRenderer: customComponentRenderer, layoutEngine: layoutEngine, layout: layout)
+        }
+        
+    }
+    
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
+        lastOffset = scrollView.contentOffset.x
+    }
+    
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint,
+                                          targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        guard isSnapToCellEnabled else { return }
+        
+        let currentOffset = CGFloat(scrollView.contentOffset.x)
+        
+        if currentOffset == lastOffset {
+            return
+        }
+        
+        let lastPosition = selected
+        if currentOffset > lastOffset {
+            if lastPosition < items.count - 1 {
+                selected = lastPosition + 1 // Move to the right
+                scrollToItem(selected, animated: true)
+            }
+        } else if currentOffset < lastOffset {
+            if lastPosition >= 1 {
+                selected = lastPosition - 1
+                scrollToItem(selected, animated: true) // Move to the left
```

Comment:
I would dispatch the message explicitly by calling `dispatchOnSelectedMessage` before `scrollToItem`.